Save data and reload Test Day data across bot restarts

galgeek opened the pull request:
Hi, @whimboo!

Here's a beginning for issue #43 , saving and reloading the schedule, plus support for opt out.

I'll look at saving metrics next.

Thank you for all your help!

galgeek commented:
Hi, @whimboo and @ashughes1!

I have this update running in #testdaybotTest, and the save and reload tests are running fine for me. The bot saves the optout list, the current schedule, and the collected metrics, and reloads them after the bot restarts.

Anthony, perhaps you'll have time to take a look and let us know if there's anything vital missing?

Thank you!

whimboo commented on bot.js:
```diff
@@ -17,8 +17,8 @@ var ircServer = config.server,
-    startTime = Date.now(),
-    endTime = startTime,
+    startTime = new Date(),
+    endTime = new Date(2000),
```
Can you please describe this change?

galgeek replied:
Date.now() returns a number, with no helpful Date methods attached.

new Date() creates a Date object. I'm initializing endTime with a guaranteed past time so that bot can check easily whether it's just been started.

whimboo commented on bot.js:
```diff
@@ -361,3 +372,92 @@ Stats.prototype.generateStats = function(metrcs, from) {
+      data = JSON.stringify(metrics);
+      break;
+    case ("optout"):
+      data = optOut.toString();
```
Please use JSON.stringify here too

whimboo commented on bot.js:
```diff
@@ -361,3 +372,92 @@ Stats.prototype.generateStats = function(metrcs, from) {
+        endTime: endTime,
+        etherpad: etherpad,
+        topic: topic,
+        topic_backup: topic_backup
```
Not all of this data is part of the schedule, e.g. topic_backup has nothing to do with it. Also I would prefer if we setup those objects already on the irc instance. So we do not have to start mangling with objects here.

galgeek replied:
To restore the state of the testdaybot after restart during a test day, we need topic_backup so that we can restore the channel topic.

whimboo replied:
But it has nothing to do with the schedule as I mentioned in my last comment. It's a property which holds a status value.

whimboo commented on bot.js:
```diff
@@ -361,3 +372,92 @@ Stats.prototype.generateStats = function(metrcs, from) {
+
+function saveData(datastore) {
+  var filename = "./data/" + datastore + ".txt",
+      data;
```
Please use var on each single line.

whimboo commented on bot.js:
```diff
@@ -361,3 +372,92 @@ Stats.prototype.generateStats = function(metrcs, from) {
   }
 };
+
+function saveData(datastore) {
```
I would expect that this method gets a filename and the json string. There should be no need for all the extra code below once our irc instance has the right objects setup.

whimboo commented on bot.js:
```diff
@@ -361,3 +372,92 @@ Stats.prototype.generateStats = function(metrcs, from) {
+            optOut = [];
+          } else {
+            optOut = data.split(",");
+          }
```
JSON.parse() would do all that automatically for you, right?

whimboo commented on bot.js:
```diff
@@ -361,3 +372,92 @@ Stats.prototype.generateStats = function(metrcs, from) {
+          endTime = new Date(schedule.endTime);
+          etherpad = schedule.etherpad;
+          topic = schedule.topic;
+          topic_backup = schedule.topic_backup;
```
I would not modify the properties directly. We should return an object instead.

whimboo commented on bot.js:
```diff
@@ -361,3 +372,92 @@ Stats.prototype.generateStats = function(metrcs, from) {
+            testDay = true;
+            timerID = setTimeout(updateTestDayData, endTime - Date.now());
+            client.send('TOPIC', channel, topic);
+          }
```
I don't see this code related to the load event. Please move it out of this method to a better place.

whimboo commented on bot.js:
```diff
@@ -361,3 +372,92 @@ Stats.prototype.generateStats = function(metrcs, from) {
+  });
+}
+
+function readData(datastore) {
```
Ideally I would expect this method to accept the datastore type and return an object for further processing.

whimboo commented on bot.js:
```diff
@@ -361,3 +372,92 @@ Stats.prototype.generateStats = function(metrcs, from) {
+  });
+}
+
+function readRestartData () {
```
Maybe call this method initialize, and move it up to the top.

whimboo commented on bot.js:
```diff
@@ -361,3 +372,92 @@ Stats.prototype.generateStats = function(metrcs, from) {
+  var filename = "./data/" + datastore + ".txt",
+  data;
+
+  fs.readFile(filename, "utf8", function(err, data){
```
nit: please keep blanks in mind.

galgeek commented:
Hello, @whimboo and @ashughes1!

I’ve updated per Henrik's Friday feedback—thank you, Henrik! I've added just a couple of line notes where you had questions.

This code is running now in channel #testdaybotTest. Note that without op privileges, there'll be no topic changes. I was able to verify the topic change code works earlier. 

Send :help to get a list of bot commands via private message.

whimboo commented on bot.js:
```diff
@@ -11,23 +11,27 @@ var ircServer = config.server,
+      etherpad: "",
+      topic: "",
+      topic_backup: "",
+      advertisement: config.advertisement,
```
You really want to separate out schedule and status information. That would also allow us to plan upcoming testdays while a current testday is still active. So please think about what's scheduling data, and what is current status.

galgeek replied:
I wonder if adding additional scheduling features should be a new issue. The code is pretty usable as is.

whimboo replied:
That was meant as a hint and should not have actually be done on this PR.

whimboo commented on bot.js:
```diff
@@ -11,23 +11,27 @@ var ircServer = config.server,
       hourUTC: {},
       optOutTotal: 0,
+      start: new Date(2000),
+      end: new Date(2000)
```
Maybe this block could be about the status information of a testday.

galgeek replied:
The metrics object focuses on the data we collect during a test day and want to remember for awhile, until it's reset at the beginning of a subsequent test day.

whimboo replied:
Lets leave it as is. We could pick this up later for a refactoring if necessary.

whimboo commented on bot.js:
```diff
@@ -50,40 +54,47 @@ var ircServer = config.server,
+    etherpad: testDay.etherpad,
+    topic: testDay.topic,
   };
+  saveData("metrics", JSON.stringify(metrics));
```
nit: please try to separate code blocks a bit by empty lines.

whimboo commented on bot.js:
```diff
@@ -50,40 +54,47 @@ var ircServer = config.server,
-    testDay = false;
-    client.send('TOPIC', channel, topic_backup);
+  if (testDay.end < Date.now()) {
+    testDay.inProgress = false;
```
I would call this `.active`.

whimboo commented on bot.js:
```diff
@@ -361,3 +379,55 @@ Stats.prototype.generateStats = function(metrcs, from) {
 };
+
+function restoreTestDayData() {
+  var data = readData("optOut");
```
I wonder if we should use a single status file for all the data to be saved off.

galgeek replied:
I don't think we want to use a single file for all of it. Each of the files we're saving now has its own requirements, with two of them containing personal info that we should be careful with.

The opt out data, for the bot's internal use, does maintain minimal arguably personally identifiable data, so it should be private.

The metrics data is information that Test Day organizers and perhaps others want to be able to use easily, but since it includes personally identifiable info, it shouldn't be public.

Current test day data is fine to make public, maybe with the exception of the admins.

whimboo replied:
I don't say we have to make anything public. It's only about storing the this information on disk. Which specific command actually accesses which information, is another point and has all already been implemented.

Anyway, lets get rid of this idea for now and keep what you have. At some point we have to do a refactoring of the code, but as long as this is not necessary we should not spend our time on it.

ashughes1 commented:
> Current test day data is fine to make public, maybe with the exception of the admins.

I would prefer, at least for now, if all data was only readable by the testday organizer and myself. It should be at our discretion which information is publicized. Going one step further, I would say that the bot should not share data with an organizer unless they have a registered IRC nick.

galgeek commented:
Hi, @ashughes1 and @whimboo!

Test Day bot metrics are available only to those on the Test Day admins list and logged in to a registered IRC nick, using the :stats command via private message to the bot.

When I used "Current test day data" above, I was referring to schedule, etherpad, topic, helpers, admins for an active test day. The only bit of that we're not already publishing with the bot is the admin list, and I'm not proposing to publish that, just noting that it doesn't raise personal data protection issues.

galgeek commented:
Hi, @whimboo!

I've pushed a couple of additional commits, addressing some of your feedback.

whimboo commented:
I think that looks good now. Given that I cannot kill the bot and see that everything gets restored correctly, I will trust your testing.

whimboo commented:
Commits have been squashed, the commit message updated, and the PR got merged as https://github.com/mozilla/testdaybot/commit/5ef21bad513fd5c9322d9d410a4ab4383433fc8a.
